### ISSUE TITLE
[RFC]: Encoder-Only Prefill in P/D deployment for DeepSeek V4.1 Flash

### ISSUE THREAD

**Issue by hczboy** [vllm-project/vllm#57738]
## Summary

DeepSeek-V4.1 Flash can construct the target model's reusable global KV state without executing the complete decoder stack. In a P/D deployment where Decode already rebuilds sliding-window attention (SWA) state through bounded replay, this creates an opportunity to eliminate decoder computation on Prefill and initialize DSpark entirely on Decode.

This RFC proposes:

- P processes the **entire prompt** through the causal encoder L0–L19 and the minimal L20 path needed to produce the decoder's shared Main KV and Indexer K.
- P transfers the target's required global cache state. It does not produce or transfer target SWA or DSpark caches for this handoff.
- D uses the existing cache-hit-based bounded replay boundary, recomputes the tail through the complete target model, and samples the first output token.
- D derives DSpark context KV from the replay's target auxiliary hidden states, generates its own draft candidates, and continues with normal speculative decoding.

Unlike [RFC #57383](https://github.com/vllm-project/vllm/issues/57383), this proposal does **not** reserve the last window of prompt tokens for first-time computation on D. P encodes all prompt positions, preserving the existing distinction between a cached replay window and an uncached suffix.

This is a design proposal, not an implemented or benchmarked optimization. “Encoder-only” includes the necessary boundary projections; it does not mean returning immediately after L19 without producing the decoder's global cache.

## Motivation

A conventional P worker retains the complete target model. When DSpark is enabled on P, the worker can also derive draft context KV and execute a proposal after target sampling. The example NIXL proxy forwards the original request and KV-transfer metadata to D, rather than forwarding P's sampled output and draft candidates as D's generation state.

When D subsequently performs bounded replay, it executes the target tail again and obtains the auxiliary features needed to initialize DSpark locally. Producing those features and draft caches on P is therefore avoidable if the handoff explicitly assigns their initialization to D.

The intended benefits are:

1. Remove target decoder attention/MoE computation from P, beyond the dependencies required to produce global cache state.
2. Allow P to omit decoder and DSpark weights in a later implementation phase.
3. Remove target SWA and draft cache transfers that D can rebuild.
4. Preserve the existing cache-hit-based replay model instead of introducing a new prompt-token ownership boundary.

The first implementation should prioritize correctness and a small execution change. Weight placement and GPU-count savings should be evaluated separately.

## Current Behavior and Code Baseline

The analysis below is based on vLLM commit [`41c4a3ed4e45792739bd8a5bbcd01269b06eae2c`](https://github.com/vllm-project/vllm/tree/41c4a3ed4e45792739bd8a5bbcd01269b06eae2c), specifically the GPU Model Runner V2 and NIXL pull path. It is a source-level analysis, not an end-to-end validation of every P/D configuration.

- The [example NIXL proxy](https://github.com/vllm-project/vllm/blob/41c4a3ed4e45792739bd8a5bbcd01269b06eae2c/tests/v1/kv_connector/nixl_integration/toy_proxy_server.py) asks P for one output token, extracts KV-transfer metadata, and sends the original request to D.
- The [runner](https://github.com/vllm-project/vllm/blob/41c4a3ed4e45792739bd8a5bbcd01269b06eae2c/vllm/v1/worker/gpu/model_runner.py) invokes the configured speculator after sampling and before connector post-forward processing.
- [DSpark context precompute](https://github.com/vllm-project/vllm/blob/41c4a3ed4e45792739bd8a5bbcd01269b06eae2c/vllm/models/deepseek_v41/nvidia/dspark.py) projects target auxiliary features into each draft layer's window KV. The configured target layer IDs must retain their V4.1 attention-input semantics; they are not interchangeable with final normalized target output.
- [NIXL block exchange](https://github.com/vllm-project/vllm/blob/41c4a3ed4e45792739bd8a5bbcd01269b06eae2c/vllm/distributed/kv_transfer/kv_connector/v1/nixl/base_scheduler.py) selects transferable groups and clips sliding-window groups to their tail blocks. It does not automatically omit a group merely because `prefix_cacheable=False`.
- The [scheduler](https://github.com/vllm-project/vllm/blob/41c4a3ed4e45792739bd8a5bbcd01269b06eae2c/vllm/v1/core/sched/scheduler.py) rewinds a loaded prefix by the replay window. [V4.1 model state](https://github.com/vllm-project/vllm/blob/41c4a3ed4e45792739bd8a5bbcd01269b06eae2c/vllm/models/deepseek_v41/nvidia/model_state.py) masks replay writes to prefix-cacheable groups while rebuilding SWA.

Consequently, removing P's decoder forward requires an explicit new handoff contract. It is not sufficient to skip layers while advertising their uninitialized caches as valid.

### Model placement and execution

Use zero-based target layer numbering throughout.

| Component | P | D |
| --- | --- | --- |
| Embeddings and causal encoder L0–L19 | Run for all prompt positions, subject to existing valid prefix reuse | Run for replay and uncached suffix positions, then generated tokens |
| L20 global KV/Indexer producer path | Run for all positions requiring encoding | Preserve loaded entries during replay; produce new entries when appropriate |
| Remaining L20 computation and L21–L39 | Skip | Execute normally |
| Target SWA | Maintain encoder-local SWA needed for P computation; do not transfer it under this mode | Allocate and rebuild locally |
| Final normalization, LM head, output sampling | Not needed for this handoff | Generate the first output and subsequent outputs |
| DSpark auxiliary features and context KV | Do not produce | Produce from local target execution |
| DSpark proposals and target verification | Do not run | Run normally after initialization |

The L20 producer path must preserve every upstream operation needed for correct global cache bytes, including relevant mHC/normalization, projections, positional transforms, compression or packing, quantization metadata, and cache writes. It must not be approximated as an isolated linear projection.

P still needs its encoder SWA during encoding. “Do not transfer SWA” does not mean “do not compute encoder SWA.”

### Transfer contract

The new mode should explicitly identify the global cache groups produced by P and the locally reconstructed groups on D.

| State | Handoff policy |
| --- | --- |
| Required target global Main KV and Indexer cache state | Transfer, including the representation's required scales and metadata |
| Required boundary/compression continuation state | Preserve or reconstruct under an explicitly validated boundary rule |
| Target SWA | D-local reconstruction |
| DSpark context and query caches | D-local initialization |
| Target auxiliary hidden states | Not transferred |
| P output token, logits, draft IDs, draft probabilities | Not produced for handoff |
| Prompt, request configuration, cache coverage, transfer identity | Retained in the control path |

Declare a prefix available only after every required global group covers that prefix. A missing required global group remains a load failure. Missing SWA/draft state is expected under this mode and must not be mistaken for such a failure.

Do not infer the transfer set solely from `prefix_cacheable`. Use an explicit model-supported group selection, with matching layer identities, layouts, dtypes, quantization metadata, coverage, and TP/PP mapping on both sides.

### Replay boundary

Let:

- `N` be the prompt length;
- `H` be the actual prefix length adopted by D after connector limits and block alignment;
- `W` be the validated replay window, initially 128 for the supported configuration.

For an adopted hit with `H > W`:

```text
replay_start = H - W
cached replay positions = [H - W, H)
uncached suffix positions = [H, N)
D target positions to execute = [H - W, N)
```

Thus D processes `W + (N - H)` prompt positions in total, possibly across scheduler steps. It is exactly `W` only when `H == N`.

For hits too short to amortize replay, preserve the existing full-recompute behavior. For short prompts, D can run a cold prefill; bypassing P for such requests is an optional routing optimization, not a prerequisite.

During replay, preserve loaded global cache entries and prevent duplicate compression-state advancement. Newly computed suffix positions must still populate their required global caches. Block and compression boundaries remain authoritative.

### Request flow

1. The router sends P an explicit cache-production request for the full prompt.
2. P executes the encoder and required boundary producer path.
3. P reports cache readiness and transferable coverage after the relevant device work is complete. No actual generated token is required to complete this operation.
4. D receives the original request, allocates local state, and loads the declared global groups.
5. After load completion, D applies the existing cache-hit replay boundary.
6. D executes the target replay and any uncached suffix, rebuilding target SWA and exporting DSpark auxiliary features.
7. D samples the first output token when prompt processing completes.
8. D initializes DSpark context KV from valid target features before its first real proposal, then performs ordinary draft/verify iterations.

### Example

Assume a 1,024-token prompt, an adopted 1,024-token global-cache hit, a 128-token window, and three draft candidates per proposal.

| Step | Work |
| --- | --- |
| P | Encode positions `[0, 1024)` and produce required global caches; no decoder-tail forward or DSpark proposal |
| Transfer | Load global cache groups into D; do not load target SWA or draft groups |
| D replay | Execute target positions `[896, 1024)`, preserving loaded global entries and rebuilding local SWA |
| D sampling | Sample the first output `y0` from the final prompt position |
| D DSpark initialization | Combine local auxiliary features and write valid context KV for each draft layer |
| D proposal | Use `y0` as anchor to propose `d1, d2, d3` |
| D verification | Run target verification on `y0, d1, d2, d3`; retain accepted output and continue |

Rejected suffix rows must not become valid draft context. Existing rejection-aware context-slot handling should remain in effect.

### DSpark coverage requirement

D-local initialization is valid only if the locally recomputed target rows cover the context that the draft attention can read. Validate this against the draft model's actual attention window and position conventions; do not assume any arbitrary DSpark checkpoint fits a 128-token replay.

Initially, support only a known compatible V4.1 target/drafter pair. A wider-context or full-attention drafter requires a larger validated reconstruction scheme or a different handoff and is outside the initial scope.

Chunking must not expose uninitialized draft context. If the current speculator would run before enough replay context exists, add a request-local initialization guard or constrain initial admission/chunking. Final draft initialization must complete before a real proposal is used for verification.

### Phase 1: Prove the execution and handoff contract

- Add an opt-in P execution mode that runs the global-cache producer path and skips decoder/output/draft work.
- Initially allow full target weight loading on P to keep weight-loader changes out of the first patch.
- Add an explicit cache-ready completion path instead of relying on `max_tokens=1` and real sampling. Propagate it through scheduler, runner, connector lifecycle, and router response handling.
- Restrict the transfer contract on both sides to required global state. Keep local D SWA/draft allocation independent from that transfer selection.
- Validate compatibility for the transferred target state even when P does not instantiate a DSpark speculator. Do not globally disable compatibility checking to permit asymmetric configurations.
- Reuse the existing replay machinery and local DSpark context precompute. Audit their initialization ordering, chunking, and write masks under this handoff.

### Phase 2: Reduce P memory residency

After correctness and performance are established, omit unused decoder, final-output, and DSpark weights on P. Preserve the original model architecture/configuration and explicitly select the execution/placement role; do not simulate this by changing the model's layer count.

Weight savings must be measured with embeddings, Engram placement, global caches, temporary buffers, CUDA graphs, and allocator headroom included. No specific single-GPU fit is claimed here.

### Initial support boundary

Start with text-only requests, NIXL pull, GPU Model Runner V2, PP=1, no prefill context parallelism or microbatching, and a fixed supported DSpark configuration. Require the replay feature to be active; fail clearly or use the complete conventional path when it is unavailable.

Prompt-logprob requests require a separate fallback because their requested outputs cannot generally be produced by an encoder-only P path and a short D replay. Extend parallelism and connector support after the initial contract is validated.

## Relationship to RFC #57383

| Aspect | RFC #57383 | This proposal |
| --- | --- | --- |
| P prompt range | Prefix `[0, N-W)` | Entire prompt `[0, N)` |
| P model path | Encoder plus necessary boundary projection | Same intended producer path |
| D first target range | Intended final window `[N-W, N)` | `[H-W, N)`, based on adopted cache coverage |
| Tail global KV | First produced by D | Already produced by P where transferred; preserved during replay |
| DSpark initialization | Needs explicit integration | Explicitly assigned to D |
| Scheduler semantics | Requires coordinating reserved-tail execution with replay | Preserves hit-based replay semantics |
| Extra P work | Avoids encoding the reserved tail | Encodes that tail as well |

If `H=N-W`, applying today's additional `W`-token rewind would start D at `N-2W`. This proposal avoids that particular mismatch by keeping full-prompt encoding and interpreting D work through the actual cache hit.

The trade-off is deliberate: retain a bounded amount of duplicate encoder work to reduce the initial change in scheduling and cache-boundary semantics. This is not a claim that every component of this RFC is already supported.

## Expected Benefits and Costs

The computation benefit depends on the baseline. Against full-prompt execution through the complete target, removing decoder work can be substantial. Against an implementation that already restricts decoder work to a trailing window, the removed work is only that decoder tail plus P-side DSpark/output work.

D still pays target replay and context-KV projection costs. If the baseline already performs the same replay, this proposal does not inherently add another replay. Any extra cost from changed draft initialization or scheduling must be measured.

Potential benefits include lower P latency, fewer P MoE/collective operations, less cache-transfer traffic, and eventually lower P weight residency. Potential costs include D admission bursts, non-uniform batches or CUDA-graph fallback, and transfer-contract complexity.

Bounded replay is not bit-equivalent to a cold full prefill. The primary comparison should therefore be against the same replay-enabled P/D baseline; a cold/full-state path remains a secondary quality reference.

## Validation Plan

1. **Producer correctness:** compare required global cache contents and coverage with the conventional P path under identical numerical settings, including quantization scales and boundary state.
2. **Transfer correctness:** confirm that no SWA/draft groups are advertised or loaded in this mode, while every required global group is covered. Exercise block-aligned and unaligned prompt lengths.
3. **Replay correctness:** verify global-write masking, suffix writes, SWA reconstruction, short hits, load failure, cancellation, and preemption recovery.
4. **DSpark initialization:** verify auxiliary-feature selection, context coverage, initialization ordering, accepted/rejected rows, and no reads from uninitialized draft positions. Compare draft acceptance and quality against the replay-enabled baseline.
5. **Performance:** measure P latency and throughput, end-to-end TTFT, D P50/P95/P99 ITL, transferred bytes by group, draft acceptance length, and P/D memory residency. Sweep prompt length, arrival rate, and concurrent decodes.

A successful implementation must demonstrate valid state ownership and no quality regression beyond the chosen replay baseline before claims about replica sizing or throughput gains.

## Alternatives Considered

- **Keep full P execution and transfer DSpark KV:** smallest execution change when the existing symmetric path works, but retains deep target computation and weights on P. D replay may update the same draft context again.
- **Transfer target auxiliary hidden states:** still requires deep target execution on P and introduces an activation handoff; it does not achieve the main P-side savings.
- **Implement RFC #57383 directly:** potentially avoids the duplicate encoder tail, but adds reserved-tail and replay-boundary integration to the first implementation.
- **Transfer all SWA state and avoid replay:** a different latency/bandwidth trade-off that retains the P decoder state-production dependency.

## Open Questions / Feedback Requested

- What is the smallest reusable interface for cache-only request completion without a generated token?
- Should transferable-group selection be expressed as a model-specific handoff profile or a broader connector capability?
- How should compatibility checks admit P-without-DSpark while preserving strict validation of transferred target state?
- Which existing DSpark initialization guards are sufficient for chunked replay, and which need extension?
- Can weight omission be isolated from the initial execution patch without retaining unwanted cache allocations?

## Status and Ownership

Draft proposal. No implementation, benchmark results, or implementation ownership is asserted. This RFC is intended to solicit feedback on the execution boundary, cache-transfer contract, and DSpark initialization requirements.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by wangyicong52** [comment]
Thanks @hczboy for writing up this RFC. The full-prompt approach seems easier to implement than the reserved-tail split in #57383 because it can reuse the existing cache-hit replay path. For long prompts, P’s additional work is limited to one window, so I suspect the performance difference may be small. I haven’t measured the two designs directly.

Since the initial path in your RFC is NIXL, I put together a Mooncake version in #57872. It has preliminary H20 correctness and latency results that may be useful as a reference, though the PR still needs more work and validation.

If you plan to implement the NIXL path, I’d be happy to coordinate and stack the Mooncake integration on it where the shared pieces align. If that isn’t in your plans, I can look into NIXL support later as well. 

Thanks again for sharing the idea. Happy to work together to move this feature forward.

**Comment by hczboy** [comment]
@wangyicong52 ， thanks for your work on moocake integration, please also feel free to go ahead with NIXL support.

**Comment by wangyicong52** [comment]
Thanks @hczboy, I appreciate the encouragement and the RFC. I’ll focus on validating and refining the Mooncake path in #57872 first, then look into NIXL support as a follow-up. I’d like to reuse the common execution and cache-handoff logic where it makes sense, and I’ll link any NIXL work back here.
